tests/unit/models/test_email_address.py

```diff
@@ -0,0 +1,612 @@
+
+
+def test_email_hash_stability():
+    """Safety test to ensure email_blakeb160_hash doesn't change spec"""
+    ehash = email_blake2b160_hash
+    assert ehash('example@example.com') == hash_map['example@example.com']
+    assert ehash('EXAMPLE@EXAMPLE.COM') != hash_map['example@example.com']
+    assert ehash('example@gmail.com') == hash_map['example@gmail.com']
+    assert ehash('example@googlemail.com') == hash_map['example@googlemail.com']
+
+
+def test_canonical_email_representation():
+    """Test canonical email representation"""
+    cemail = canonical_email_representation
+    assert cemail('example@example.com') == ['example@example.com']
+    assert cemail('EXAMPLE@EXAMPLE.COM') == ['example@example.com']
+    assert cemail('example@googlemail.com') == [
+        'example@gmail.com',
+        'example@googlemail.com',
+    ]
+    assert cemail('example+extra@example.com') == ['example@example.com']
+    assert cemail('exam.pl.e@gmail.com') == [
+        'example@gmail.com',
+        'exam.pl.e@gmail.com',
+    ]
+    assert cemail('exam.pl.e+extra@gmail.com') == [
+        'example@gmail.com',
+        'exam.pl.e@gmail.com',
+    ]
+    assert cemail('exam.pl.e@googlemail.com') == [
+        'example@gmail.com',
+        'exam.pl.e@googlemail.com',
+    ]
+    assert cemail('exam.pl.e+extra@googlemail.com') == [
+        'example@gmail.com',
+        'exam.pl.e@googlemail.com',
+    ]
+    with pytest.raises(ValueError):
+        cemail('')
+    with pytest.raises(ValueError):
+        cemail('invalid')
+
+
+def test_email_address_init():
+    """EmailAddress instances can be created using a string email address"""
+
+    # Ordinary use constructor passes without incident
+    ea1 = EmailAddress('example@example.com')
+    assert ea1.email == 'example@example.com'
+    assert ea1.email_lower == 'example@example.com'
+    assert ea1.blake2b160 == hash_map['example@example.com']
+    assert ea1.email_canonical == 'example@example.com'
+    assert ea1.blake2b160_canonical == hash_map['example@example.com']
+    assert str(ea1) == 'example@example.com'
+    assert repr(ea1) == "EmailAddress('example@example.com')"
+    # Public hash (for URLs)
+    assert ea1.email_hash == '2EGz72jxcsYjvXxF7r5rqfAgikor'
+
+    # Case is preserved but disregarded for hashes
+    ea2 = EmailAddress('Example@example.com')
+    assert ea2.email == 'Example@example.com'
+    assert ea2.email_lower == 'example@example.com'
+    assert ea2.blake2b160 == hash_map['example@example.com']
+    assert ea1.email_canonical == 'example@example.com'
+    assert ea2.blake2b160_canonical == hash_map['example@example.com']
+    assert str(ea2) == 'Example@example.com'
+    assert repr(ea2) == "EmailAddress('Example@example.com')"
+
+    # Canonical representation's hash can be distinct from regular hash
+    ea3 = EmailAddress('Example+Extra@example.com')
+    assert ea3.email == 'Example+Extra@example.com'
```

Comment:
Why are we not using lower case by default? Email isn't case-sensitive, is it?

Reply:
Email is technically case sensitive, although in practice it isn't. Our implementation preserves the user's choice of case but enforces uniqueness against the lowercase representation.